api/path_node_userinput.go

```diff
@@ -115,3 +142,148 @@ func DeleteNodeUserInput(errorhandler DeviceErrorHandler, db *bolt.DB,
+	serviceUrl := userInput.ServiceUrl
+	serviceVersionRange := userInput.ServiceVersionRange
+	serviceArch := userInput.ServiceArch
+
+	// The versionRange field is checked for valid characters by the Version_Expression_Factory, it has a very
+	// specific syntax and allows a subset of normally valid characters.
+
+	// Use a default sensor version that allows all version if not specified.
+	if &userInput.ServiceVersionRange == nil || userInput.ServiceVersionRange == "" {
+		def := "0.0.0"
+		serviceVersionRange = def
+	}
+
+	// Convert the sensor version to a version expression.
+	vExp, err := semanticversion.Version_Expression_Factory(serviceVersionRange)
+	var errorString string
+	if err != nil {
+		errorString = fmt.Sprintf("versionRange %v cannot be converted to a version expression, error %v", userInput.ServiceVersionRange, err)
+		return false, errors.New(errorString)
+	}
+
+	// service exist
+	var sdef *exchange.ServiceDefinition
+	sdef, _, err = getService(serviceUrl, serviceOrg, vExp.Get_expression(), serviceArch)
+	if sdef == nil {
+		errorString = fmt.Sprintf("Service not exist for org: %v, url: %v, version: %v, arch: %v, get error: %v \n", serviceOrg, serviceUrl, vExp.Get_expression, serviceArch, err)
+		return false, errors.New(errorString)
+	} else if err != nil {
+		errorString = fmt.Sprintf("Error from get exchange service: %v \n", err)
+		return false, errors.New(errorString)
+	}
+
+	// compare ServiceDefinition.Userinput (array) with userInput.Inputs (array)
+	serviceUserInputs := sdef.UserInputs
+	policyUserInputs := userInput.Inputs
+
+	serviceUserinputsMap := make(map[string]exchange.UserInput)
+	var serviceInputName string
+	var serviceInput exchange.UserInput
+
+	for _, serviceInput = range serviceUserInputs {
+		serviceInputName = serviceInput.Name
+		serviceUserinputsMap[serviceInputName] = serviceInput
+	}
+
+	// the variables in userInput.Inputs are defined in the service, if not, return true with warning message, check values are correct types
+	var policyInputName string
+	var policyInputValue interface{}
+	var ok bool
+	var inputNameNotDefinedInService []string
+
+	for _, policyInput := range policyUserInputs {
+		policyInputName = policyInput.Name
+		policyInputValue = policyInput.Value
+
+		serviceInput, ok = serviceUserinputsMap[policyInputName]
+		if !ok {
+			// give back a warning with this errorString
+			inputNameNotDefinedInService = append(inputNameNotDefinedInService, policyInputName)
+		} else {
+			typeValidated, err := isCorrectType(policyInputValue, serviceInput.Type)
+			if !typeValidated {
+				return false, err
+			}
+		}
+
+	}
+
+	if len(inputNameNotDefinedInService) != 0 {
+		errorString = fmt.Sprintf("The following variables are not defined in userInput for service in its highest version: %v \n", inputNameNotDefinedInService)
+		return true, errors.New(errorString)
+	}
+
+	return true, nil
+}
+
+func isCorrectType(policyInputValue interface{}, serviceInputType string) (bool, error) {
```

Comment:
We support "list of strings" as a type also

Reply:
Added check for list of strings